llvm/include/llvm/ADT/ArrayRef.h

```diff
@@ -113,7 +115,8 @@ namespace llvm {
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Winit-list-lifetime"
 #endif
-    constexpr /*implicit*/ ArrayRef(const std::initializer_list<T> &Vec)
+    constexpr /*implicit*/ ArrayRef(
+        const std::initializer_list<T> &Vec LLVM_LIFETIME_BOUND)
```

Comment:
Hmm, actually, I am not sure if I'm super confident about this one.

Consider:
```
ArrayRef<int> ref({1, 2, 3, 4});
use(ref);
```

What would happen here? I wonder if this is actually safe despite the initializer list itself was a temporary and already destroyed when I invoke `use`.

Reply:
Actually, I am surprised we take an initializer list by const-ref here. I wonder if the right solution here is to take it by value and keep the annotation. I think lifetime annotations might work out just right after taking it by value.

Reply:
> Consider:

> ArrayRef<int> ref({1, 2, 3, 4});
> use(ref);


> Hmm, actually, I am not sure if I'm super confident about this one.
> 
> Consider:
> 
> ```
> ArrayRef<int> ref({1, 2, 3, 4});
> use(ref);
> ```
> 
> What would happen here? I wonder if this is actually safe despite the initializer list itself was a temporary and already destroyed when I invoke `use`.

This is actually a use-after-free, the backing array of the initializer_list is destroyed after the ref declaration statement, see https://godbolt.org/z/fYW7Tqcfq.

> The backing array has the same lifetime as any other [temporary object](https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime), except that initializing an [std::initializer_list](https://en.cppreference.com/w/cpp/utility/initializer_list) object from the backing array extends the lifetime of the array exactly like [binding a reference to a temporary](https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary).

Reply:
> Actually, I am surprised we take an initializer list by const-ref here. 

I think it doesn't matter much, I agree that in general `std::initializer_list` is passed by value.

Reply:
Ah, I see! Thanks! I have not realized the backing array is also a temporary with it's lifetime extended in some scenarios.

Reply:
We have a tricky case with an empty initializer_list:

```
ArrayRef<int> ref({}); 
```

clang still diagnoses it as dangling reference, but the code doesn't have any dangling references. I think it is ok, this is similar to `make_unique<pointer>(nullptr).get()` case (https://github.com/llvm/llvm-project/issues/97471).

Reply:
Now that the patch to remove the reference from the initialiser list, this should no longer warn, right?

Reply:
> Now that the patch to remove the reference from the initialiser list landed, this should no longer warn, right?

Ah, you're right. For initializer_list, we actually track its backing array, for the empty case `{}`, there is no `MaterializeTemporaryExpr` inside.